src/common/delivery/deploy/pcf/CloudFoundryApi.ts

```diff
@@ -0,0 +1,226 @@
+    const cf = {
+        api_url: cfi.api,
+        token_type: token.token_type,
+        token: token.access_token,
+        apps: new (require("cf-client")).Apps(cfi.api),
+        domains: new (require("cf-client")).Domains(cfi.api),
+        spaces: new (require("cf-client")).Spaces(cfi.api),
+        serviceBindings: new (require("cf-client")).ServiceBindings(cfi.api),
+        userProvidedServices: new (require("cf-client")).UserProvidedServices(cfi.api),
+        routes: new (require("cf-client")).Routes(cfi.api),
+    };
+    cf.apps.setToken(token);
+    cf.domains.setToken(token);
+    cf.spaces.setToken(token);
+    cf.serviceBindings.setToken(token);
+    cf.userProvidedServices.setToken(token);
+    cf.routes.setToken(token);
+    return cf;
+}
+
+export class CloudFoundryApi {
+
+    private authHeader;
+    private jsonContentHeader = {
+        "Content-Type": "application/json"
+    };
+
+    constructor(private cf: CloudFoundryClientV2,
+                      private retryInterval: number = 1000) {
+        this.authHeader = {
+            Authorization: `${cf.token_type} ${cf.token}`
+        };
+    }
+
+    private async retryUntilCondition<T>(action: () => T, condition: (T) => boolean): Promise<T> {
+        let result = await action();
+        if (condition(result)) {
+            return result
+        }
+        await new Promise( res => setTimeout(res, this.retryInterval));
+        return this.retryUntilCondition(action, condition);
+    }
+
+    public stopApp(app_guid: string): Promise<AxiosResponse<any>> {
+        return axios.post(
+            `${this.cf.api_url}/v3/apps/${app_guid}/actions/stop`,
+            undefined,
+            { headers: this.authHeader}
+        );
+    }
+
+    public startApp(app_guid: string): Promise<AxiosResponse<any>> {
+        return axios.post(
+            `${this.cf.api_url}/v3/apps/${app_guid}/actions/start`,
+            undefined,
+            { headers: this.authHeader}
+        );
+    }
+
+    public async getApp(space_guid: string, app_name: string): Promise<any> {
+        const apps = await this.cf.spaces.getSpaceApps(space_guid, {
+            q: `name:${app_name}`
+        })
+        return _.head(apps.resources);
+    }
+
+    public async createApp(space_guid: string, app_name: string): Promise<any> {
+        const existing_app = await this.getApp(space_guid, app_name);
+        return existing_app ? Promise.resolve(existing_app) : this.cf.apps.add({
+            name: app_name,
+            space_guid: space_guid,
+        });
+    }
+
+    public deleteApp(app_guid: string): Promise<any> {
+        return axios.delete(
+            `${this.cf.api_url}/v3/apps/${app_guid}`,
+            { headers: this.authHeader}
+        );
+    }
+
+    public async uploadPackage(app_guid: string, packageFile: ReadStream): Promise<AxiosResponse<any>> {
+        const packageCreateResult = await axios.post(`${this.cf.api_url}/v3/packages`, {
```

Comment:
Should we use retry? As in various other places?

`import { doWithRetry } from "@atomist/automation-client/util/retry";`